File: backend/notifications/src/main/java/com/oxygensend/notifications/infrastructure/authentication/HashingAuthentication.java
package com.oxygensend.notifications.infrastructure.authentication;

import com.oxygensend.notifications.context.authentication.AuthException;
import com.oxygensend.notifications.context.authentication.Authentication;
import org.springframework.stereotype.Component;

import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HexFormat;
import java.util.Map;

@Component
public class HashingAuthentication implements Authentication {

    private static final Charset ENCODING = StandardCharsets.UTF_8;

    @Override
    public void authenticate(Map<String, String> params) throws AuthException {
        var plainPassword = params.get("password");
        if (plainPassword == null) {
            throw new IllegalArgumentException("Password is required");
        }
        var hashedPassword = params.get("hashedPassword");
        if (hashedPassword == null) {
            throw new IllegalArgumentException("Hashed password is required");
        }

        byte[] hash = calculateHash(plainPassword);
        byte[] hashPassword = HexFormat.of().parseHex(hashedPassword);
        if (!MessageDigest.isEqual(hash, hashPassword)) {
            throw new AuthException("Invalid password");
        }
    }

    private byte[] calculateHash(String plainPassword) throws AuthException {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] passwordBytes = plainPassword.getBytes(ENCODING);
            byte[] first = md.digest(passwordBytes);
            return md.digest(first);
        } catch (NoSuchAlgorithmException e) {
            throw new AuthException(e.getMessage());
        }
    }
}

File: backend/notifications/src/main/kotlin/com/oxygensend/notifications/infrastructure/authentication/HashingAuthentication.kt
package com.oxygensend.notifications.infrastructure.authentication

import com.oxygensend.notifications.context.authentication.AuthException
import com.oxygensend.notifications.context.authentication.Authentication
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.*

@Component
internal class HashingAuthentication : Authentication {
    @Throws(AuthException::class)
    override fun authenticate(params: Map<String, String?>) {
        val plainPassword = params["password"] ?: throw IllegalArgumentException("Password is required")
        val hashedPassword = params["hashedPassword"]
            ?: throw IllegalArgumentException("Hashed password is required")
        val hash = calculateHash(plainPassword)
        val hashPassword = HexFormat.of().parseHex(hashedPassword)
        if (!MessageDigest.isEqual(hash, hashPassword)) {
            throw AuthException("Invalid password")
        }
    }

    @Throws(AuthException::class)
    private fun calculateHash(plainPassword: String): ByteArray {
        return try {
            val md = MessageDigest.getInstance("SHA-256")
            val passwordBytes = plainPassword.toByteArray(ENCODING)
            val first = md.digest(passwordBytes)
            md.digest(first)
        } catch (e: NoSuchAlgorithmException) {
            throw AuthException(e.message)
        }
    }

    companion object {
        private val ENCODING = StandardCharsets.UTF_8
    }
}
